### ISSUE TITLE
Quadractic complexity in BarycentricMapping

### ISSUE THREAD

**Issue by damienmarchal** [sofa-framework/sofa#784]
Hi all,

We are using bigger and bigger mesh at defrost and noticed that the init time of BarycentricMapping<Tetrahedron> is growing a lot. 

Eulalie pointed out that this is because there is a imbricated for() loop to find closest tetrahedron for each of the points in the target geometry. 

At line 865 in BarycentricMapping.inl in BarycentricMapperTetrahedronSetTopology
```cpp
  ....
  for ( unsigned int i=0; i<out.size(); i++ )
    {
        sofa::defaulttype::Vec3d pos = Out::getCPos(out[i]);
        sofa::defaulttype::Vector3 coefs;
        int index = -1;
        double distance = 1e10;
        for ( unsigned int t = 0; t < tetrahedra.size(); t++ )
        {
            sofa::defaulttype::Vec3d v = bases[t] * ( pos - in[tetrahedra[t][0]] );
            double d = std::max ( std::max ( -v[0],-v[1] ),std::max ( -v[2],v[0]+v[1]+v[2]-1 ) );
            if ( d>0 ) d = ( pos-centers[t] ).norm2();
            if ( d<distance ) { coefs = v; distance = d; index = t; }
        }
        if ( distance>0 )
        {
            ++outside;
        }
        addPointInTetra ( index, coefs.ptr() );
    }
```

Adding a geometric acceleration structure (aabb, sweep & prune, voxel or spatial hashing would help in keeping to a n.log.n complexity and reduce a lot the computation time. 
______________________________________________________

Suggested labels:

**Comment by damienmarchal** [comment]
I personally really like to use spatial hashing for this kind of things. 
http://matthias-mueller-fischer.ch/publications/tetraederCollision.pdf

**Comment by damienmarchal** [comment]
is the PR will be merged soon ?

**Comment by epernod** [comment]
@damienmarchal @EulalieCoevoet 
Do you confirm this problem has been solved in PR #798 and we can close this issue?

**Comment by EulalieCoevoet** [comment]
Yes
